Update System Docs Page Update (#1486)
LLNL/benchpark: docs/update-system-config.rst

>  Updating a System

Reviewer: ```suggestion
 Updating the System Software Specification
```

###################

> These steps show how to use ``benchpark system external`` to compare the

Reviewer: ```suggestion
These steps show how to use ``benchpark system external`` to compare and update the
```

external packages defined by a Benchpark system against packages detected on

> the current machine.

Reviewer: ```suggestion
the current machine. For adding a new system, please see :ref:`add-a-system-config`
```

***************************

>  Check an Existing System

Reviewer: ```suggestion
 1. Check an Existing System
```

***************************

Run ``benchpark system external`` with the same system variants that would normally be

> passed to ``benchpar system init``. For example, to check the Tioga configuration of

Reviewer: ```suggestion
passed to ``benchpark system init``. For example, to check the Tioga configuration of
```

``llnl-elcapitan``:

::

    benchpark system external llnl-elcapitan cluster=tioga

Benchpark processes the supplied system specification, collects the expected
package configuration from the system definition, and runs Spack external

> detection for those packages. The detected configuration is then compared

Reviewer: ```suggestion
Benchpark collects the current system software specification, and runs ``spack external find`` to detect new versions for each package.
The detected configuration is then compared
```

> against the configuration defined by Benchpark.

Reviewer: ```suggestion
against the current configuration defined by Benchpark.
```

To generate version-update proposals for the system
definition, use ``--propose``:

::

    benchpark system external llnl-elcapitan cluster=tioga --propose

> Additional system variants can be supplied in the same way. For example:

Reviewer: ```suggestion
Additional system variants can be specified in the same way. For example:
```

::

>     ./bin/benchpark system external llnl-elcapitan cluster=tioga --propose

Reviewer: ```suggestion
    benchpark system external llnl-elcapitan cluster=tioga --propose
```

Reviewer: Also the line above this says "additional variants can be supplied", but this line is identical to the previous benchpark system external line.

>     ./bin/benchpark system external llnl-elcapitan cluster=tioga --propose

Reviewer: More detail would be nice after this like

```
This proposed changes will be written to standard output. If you want to save these changes to a file, proceed to the next step.
```

************************

>  Save Proposed Changes

Reviewer: ```suggestion
 2. Save Proposed Changes
```

************************

Write the proposed changes as a unified diff with ``--patch-file``:

::

    benchpark system external llnl-elcapitan cluster=tioga --patch-file=tioga-externals.patch

Alternatively, write a complete proposed copy of the updated ``system.py``:

::

>     benchpark system external llnl-elcapitan cluster=tioga --write-proposed-system=system.py.proposed

Reviewer: So `--patch-file` creates a git patch file, where `--write-proposed-system` creates a patch file and applies it in one step and saves that to a new copy of my `system.py`, named `system.py.proposed`? Can you write and explain in this detail, because this is the type of question people will have.

These options generate proposed changes without modifying the existing
Benchpark system definition.

*********************

>  Apply Local Changes

Reviewer: ```suggestion
 3. Apply Local Changes
```

*********************

After reviewing the proposed updates, use ``--apply`` to apply changes
directly to the system's ``system.py``:

::

>     benchpark system external llnl-elcapitan cluster=tioga --apply

Reviewer: Does `--apply` apply the patch file I just made with `--patch-file=tioga-externals.patch`? What if I have generated multiple different patch files with different names? If it creates a new file and applies it, it seems redundant to `--write-proposed-system`.

If I am mistaken about the redundancy, then the difference between the "apply" and "save" sections should be elaborated.

Benchpark performs the same external detection and comparison before updating
the system definition. Only changes that can be matched and applied
unambiguously are written automatically. Updates that require manual review are
reported without being applied.

The resulting changes can be inspected with Git before committing:

::

    git diff -- systems/llnl-elcapitan/system.py

************************

>  Create a Pull Request

Reviewer: ```suggestion
 4. Create a Pull Request
```

************************

To apply the detected changes and create a pull request containing the update,
use ``--apply`` together with ``--pr``:

::

    benchpark system external llnl-elcapitan cluster=tioga --apply --pr

Benchpark first applies the safe updates to the system definition. It then
creates a Git branch for the update, commits the modified ``system.py``, pushes
the branch to the configured remote, and opens a pull request against the

> Benchpark repository.

Reviewer: Unless this is already implemented, I think it would be fine to omit this feature and expect the used can create a git branch and make a pull request. IMO.
